File: Data Structures/#02.Linear Data Structurs - List And DS Complexity Exercises/Java/src/T08_Double_Linked_List/DoublyLinkedList.java
package T08_Double_Linked_List;

import java.util.Iterator;
import java.util.function.Consumer;

public class DoublyLinkedList<E> implements Iterable<E> {

    private int size;
    private Node<E> head;
    private Node<E> tail;

    public int size() {
        return this.size;
    }

    private void setSize(int size) {
        this.size = size;
    }

    public void addFirst(E element) {

        if (this.size == 0) {
            Node<E> node = new Node<>(element, null, null);
            this.head = node;
            this.tail = node;
            this.size++;
            return;
        }

        Node<E> node = new Node<>(element, head, null);
        this.head = node;
        this.size++;

    }

    public void addLast(E element) {

        if (this.size == 0) {
            Node<E> node = new Node<>(element, null, null);
            this.head = node;
            this.tail = node;
            this.size++;
            return;
        }

        Node<E> node = new Node<>(element, null, this.tail);
        this.tail = node;
        this.size++;
    }

    public E removeFirst() {

        Node<E> removedNode = this.head.getNext();
        this.head = this.head.getNext();
        this.size--;

        return removedNode.getValue();
    }

    public E removeLast() {
        Node<E> removedNode = this.tail.getPrevious();
        this.tail = this.tail.getPrevious();
        this.size--;

        return removedNode.getValue();
    }

    public E[] toArray() {
        return null;
    }

    @Override
    public Iterator<E> iterator() {
        return null;
    }

    @Override
    public void forEach(Consumer<? super E> action) {

    }

}

File: Data Structures/#02.Linear Data Structurs - List And DS Complexity Exercises/C#/List And DS Complexity Exercises/T08_Doubly_Linked_List/DoublyLinkedList.cs
using System;
using System.Collections;
using System.Collections.Generic;

public class DoublyLinkedList<T> : IEnumerable<T>
{
    private Node<T> head;
    private Node<T> tail;
    private int count;

    public int Count { get => count; set => count = value; }

    public void AddFirst(T element)
    {
        if (this.count == 0)
        {
            Node<T> firstNode = new Node<T>(element, null, null);
            this.head = firstNode;
            this.tail = firstNode;
            this.count++;
            return;
        }

        Node<T> newNode = new Node<T>(element, this.head, null);
        this.head.PreviousNode = newNode;
        this.head = newNode;
        this.count++;

    }

    public void AddLast(T element)
    {
        if (this.count == 0)
        {
            Node<T> firstNode = new Node<T>(element, null, null);
            this.head = firstNode;
            this.tail = firstNode;
            this.count++;
            return;
        }

        Node<T> newNode = new Node<T>(element, null, this.tail);
        this.tail.NextNode = newNode;
        this.tail = newNode;
        this.count++;

    }

    public T RemoveFirst()
    {

        if (this.count == 0)
        {
            throw new InvalidOperationException();
        }

        T element = this.head.Value;

        if (this.count == 1)
        {
            this.head = null;
            this.tail = null;
            this.count = 0;
            return element;
        }

        Node<T> newNode = this.head.NextNode;
        this.head.PreviousNode = null;
        newNode.PreviousNode = null;
        this.head = newNode;
        this.count--;

        return element;
    }

    public T RemoveLast()
    {
        if (this.count == 0)
        {
            throw new InvalidOperationException();
        }

        T element = this.head.Value;

        if (this.count == 1)
        {
            this.head = null;
            this.tail = null;
            this.count = 0;
            return element;
        }

        Node<T> newNode = this.tail.PreviousNode;
        this.tail.NextNode = null;
        newNode.NextNode = null;
        this.tail = newNode;
        this.count--;

        return element;
    }

    public void ForEach(Action<T> action)
    {
        Node<T> currentNode = this.head;

        while (currentNode != null)
        {
            action.Invoke(currentNode.Value);
            currentNode = currentNode.NextNode;
        }
    }

    public IEnumerator<T> GetEnumerator()
    {
        List<T> result = new List<T>();
        result.AddRange(this.ToArray());
        return result.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return this.ToArray().GetEnumerator();
    }

    public T[] ToArray()
    {

        T[] arr = new T[this.count];
        Node<T> currentNode = this.head;

        int counter = 0;
        while (currentNode != null)
        {
            arr[counter] = currentNode.Value;
            currentNode = currentNode.NextNode;
            counter++;
        }

        return arr;
    }
}
